packages/core/admin/server/services/__tests__/api-token.test.js

```diff
@@ -254,15 +288,209 @@ describe('API Token', () => {
       };
 
       const res = await apiTokenService.update(id, attributes);
-
+      // ensure any existing permissions have been deleted
+      expect(deleteFn).toHaveBeenCalledWith({
+        where: {
+          token: id,
+        },
+      });
       expect(update).toHaveBeenCalledWith({
-        select: ['id', 'name', 'description', 'type', 'createdAt'],
+        select: SELECT_FIELDS,
         where: { id },
         data: attributes,
+        populate: ['permissions'],
       });
       expect(res).toEqual(attributes);
     });
   });
+
+  test('Updates a custom token', async () => {
+    const id = 1;
+
+    const originalToken = {
+      id,
+      name: 'api-token_tests-name',
+      description: 'api-token_tests-description',
+      type: 'custom',
+      permissions: ['admin::subject.keepThisAction', 'admin::subject.oldAction'],
+    };
+
+    const updatedAttributes = {
+      name: 'api-token_tests-updated-name',
+      description: 'api-token_tests-description',
+      type: 'custom',
+      permissions: [
+        'admin::subject.keepThisAction',
+        'admin::subject.newAction',
+        'admin::subject.otherAction',
+      ],
+    };
+
+    const update = jest.fn(({ data }) => Promise.resolve(data));
+    const findOne = jest.fn().mockResolvedValue(omit('permissions', originalToken));
+    const deleteFn = jest.fn();
+    const create = jest.fn();
+    const load = jest
+      .fn()
+      // first call to load original permissions
+      .mockResolvedValueOnce(
+        Promise.resolve(
+          originalToken.permissions.map((p) => {
+            return {
+              action: p,
+            };
+          })
+        )
+      )
+      // second call to check new permissions
+      .mockResolvedValueOnce(
+        Promise.resolve(
+          updatedAttributes.permissions.map((p) => {
+            return {
+              action: p,
+            };
+          })
+        )
+      );
+
+    global.strapi = {
+      query() {
+        return {
+          update,
+          findOne,
+          delete: deleteFn,
+          create,
+        };
+      },
+      config: {
+        get: jest.fn(() => ''),
+      },
+      entityService: {
+        load,
+      },
+    };
+
+    const res = await apiTokenService.update(id, updatedAttributes);
+
+    expect(deleteFn).toHaveBeenCalledTimes(1);
+    // expect(deleteFn).toHaveBeenCalledWith({
```

Comment:
Forgot to remove this?

Reply:
Will discuss this issue later, lower priority performance enhancement issue